### PR TITLE
Add possibility for downloading big IMAP attachments to a file

### PR DESCRIPTION
**Pull request by disaykin**

### CODE REVIEW

File: src/low-level/imap/mailimap_types.h [hunk 1]
```diff
@@ -3341,6 +3341,8 @@ enum {
 
 typedef void mailimap_msg_att_handler(struct mailimap_msg_att * msg_att, void * context);
 
+typedef bool mailimap_msg_body_handler(const char *str, size_t len, void * context);
```
Comment (dinhviethoa) [inline src/low-level/imap/mailimap_types.h]: Could you add:
int msg_att_type
(possible values: MAILIMAP_MSG_ATT_BODY_SECTION, MAILIMAP_MSG_ATT_RFC822_TEXT, MAILIMAP_MSG_ATT_RFC822_HEADER, MAILIMAP_MSG_ATT_RFC822)
and struct mailimap_msg_att_body_section * section
to the parameters?

I think you would need to add those to mailimap_parser_context as well.

**Comment by disaykin** [comment]
Done

File: src/low-level/imap/mailimap_parser.c [hunk 2]
```diff
@@ -6017,43 +6020,51 @@ static int mailimap_literal_parse_progress(mailstream * fd, MMAPString * buffer,
...
-        goto free_literal;
-      }
-    if (progr_rate != 0) {
-      if (progr_fun != NULL) {
-        progr_fun(number, number);
+  needed = number;
+
+  uint32_t current_prog = 0;
+
+  if (left > 0) {
```
Comment (dinhviethoa) [inline src/low-level/imap/mailimap_parser.c]: Is there a way to rewrite it that would be similar when comparing with the previous code?
I makes it hard for me to figure out what is going to break here.
One thing I notice though: `progr_fun(number, number);` is not called any more.

File: src/low-level/imap/mailimap_parser.c [hunk 3]
```diff
@@ -7885,9 +7917,17 @@ mailimap_msg_att_rfc822_parse_progress(mailstream * fd, MMAPString * buffer, str
   r = mailimap_space_parse(fd, buffer, &cur_token);
   if (r != MAILIMAP_NO_ERROR)
     return r;
-  
+
+  if (parser_ctx) {
```
Comment (dinhviethoa) [inline src/low-level/imap/mailimap_parser.c]: Do you really need to check parser_ctx? It sounds like we can make assumption it's always != NULL.

Comment (dinhviethoa) [inline src/low-level/imap/mailimap_parser.c]: Let me know if I'm wrong on that one.

File: src/low-level/imap/mailimap_parser.c (hunk 2 above)
Comment (disaykin) [inline src/low-level/imap/mailimap_parser.c]: The progr_fun(number, number); is called on line 6161 or in while-block.
But I can return old logic if needed.

Comment (dinhviethoa) [inline src/low-level/imap/mailimap_parser.c]: Ah, you're right! I would be interesting in one of these options:
1/ in a version that would be easy to compare
2/ more comments in the code.
3/ more inline github comments about what you moved around.

**Comment by disaykin** [comment]
Done

**Pull request merged**